File: voice_control/bridge/clients/cpp/llm_client.cpp
#include "llm_client.hpp"
#include <iostream>

LLMClient::LLMClient(const std::string& endpoint, const std::string& authToken)
    : _context(1), _socket(_context, zmq::socket_type::req), _endpoint(endpoint), _authToken(authToken), _idCounter(0) {
}

LLMClient::~LLMClient() {
    disconnect();
}

void LLMClient::connect() {
    _socket.connect(_endpoint);
    std::cout << "[LLMClient] Connected to " << _endpoint << std::endl;
}

void LLMClient::disconnect() {
    _socket.close();
    std::cout << "[LLMClient] Disconnected" << std::endl;
}

json LLMClient::query(const std::string& content, const std::string& role) {
    json params = {
        {"content", content},
        {"role", role}
    };
    return _request("query", params);
}

json LLMClient::_request(const std::string& method, const json& params) {
    _idCounter++;
    json request = {
        {"jsonrpc", "2.0"},
        {"method", method},
        {"params", params},
        {"id", _idCounter}
    };

    if (!_authToken.empty()) {
        request["auth_token"] = _authToken;
    }

    _socket.send(zmq::buffer(request.dump()), zmq::send_flags::none);

    zmq::message_t reply;
    auto received = _socket.recv(reply, zmq::recv_flags::none);

    if (!received.has_value()) {
        throw std::runtime_error("No response from server.");
    }

    auto response = json::parse(reply.to_string());
    if (response.contains("error")) {
        throw std::runtime_error("RPC Error: " + response["error"].dump());
    }

    return response["result"];
}

// Example Usage
// You would need to link against zmq and nlohmann_json libraries.
// g++ -std=c++17 llm_client.cpp -o example -lzmq -lnlohmann_json
int main() {
    try {
        LLMClient client;
        client.connect();
        json response = client.query("Hello from C++!");
        std::cout << "Response from LLM: " << response.dump(4) << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
    }
    return 0;
}

File: voice_control/bridge/clients/python/llm_client.py
import zmq
import json


class LLMClient:
    """
    A Python client for the LLMServer.
    """

    def __init__(self, endpoint="tcp://0.0.0.0:8000", auth_token=None):
        self.endpoint = endpoint
        self.auth_token = auth_token
        self.context = zmq.Context()
        self.socket = self.context.socket(zmq.REQ)

    def connect(self):
        """Connects to the server."""
        self.socket.connect(self.endpoint)
        print(f"[LLMClient] Connected to {self.endpoint}")

    def disconnect(self):
        """Disconnects from the server."""
        self.socket.setsockopt(zmq.LINGER, 0)
        self.socket.close()
        self.context.term()
        print("[LLMClient] Disconnected")

    def _request(self, method, params):
        request = {
            "jsonrpc": "2.0",
            "method": method,
            "params": params,
            "id": 1,  # Simple ID for example
        }
        if self.auth_token:
            request["auth_token"] = self.auth_token

        self.socket.send_json(request)
        response = self.socket.recv_json()

        if "error" in response:
            raise Exception(f"RPC Error: {response['error']}")

        return response.get("result")

    def query(self, content, role="user"):
        """Sends a query to the LLM."""
        return self._request("query", {"content": content, "role": role})


if __name__ == "__main__":
    # Example usage
    client = LLMClient()
    client.connect()
    try:
        response = client.query("Hello, world!")
        print("Response from LLM:", response)
    except Exception as e:
        print(e)
    finally:
        client.disconnect()
